bindings/go/scip/parse.go

```diff
@@ -0,0 +1,149 @@
+package scip
+
+import (
+	"io"
+
+	"google.golang.org/protobuf/encoding/protowire"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// IndexVisitor is a struct of functions rather than an interface since Go
+// doesn't support adding new functions to interfaces with default
+// implementations, so adding new functions here for new fields in
+// the SCIP schema would break clients. Individual functions may be nil.
+type IndexVisitor struct {
+	VisitMetadata       func(*Metadata)
+	VisitDocument       func(*Document)
+	VisitExternalSymbol func(*SymbolInformation)
```

Comment:
I guess part of the question is, is this API fine for usage in the backend? I think it should be, as the backend can intern things when processing each Document, without having to perform any kind of ReadAll operation. Exposing an iterator style API with NextDocument() etc methods etc. is also possible, but that would be more complicated to use in Sourcegraph, with unclear benefits...

Reply:
If we can enforce an order here, metadata, then external symbols, then documents would be the best. The way we process today we build a map of external symbols that is available for each document.

Reply:
https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a2c0e6a8bf5e55227c0bca7db2e4ac0a20c1162d/-/blob/enterprise/internal/codeintel/uploads/internal/background/processor/scip.go?L30:6

Reply:
> If we can enforce an order here, metadata, then external symbols, then documents would be the best

Protobuf doesn't actually guarantee that. The tag length value representation means that fields can be interleaved.

https://protobuf.dev/programming-guides/encoding/

So the parser can't really ensure that without actually buffering the input. On the consumer side, there are a couple of options:

1. Buffer compressed versions of the documents (compressed because string interning will likely reduce memory usage), and process the external symbols as they arrive.
2. Do two traversals over the index, first for the external symbols only, and then for the documents only (or vice-versa, if needed).

Reply:
Ah ok two traversals wouldn't be bad, just set up two visitors one with a no-op for documents and one with a no-op for symbols.

Makes sense!